### PR TITLE
Feat/conformal prediction

### LINKED ISSUE #1704: Conformal Predictions
Conformal predictions could be a valuable addition to darts.
It would require some brain storming/planning of how (or if) we can integrate this into our API / extend the API.  

Some links: 
- https://valeman.medium.com/probabilistic-forecasting-with-conformal-prediction-and-neuralprophet-af9c87901d94
- https://towardsdatascience.com/time-series-forecasting-with-conformal-prediction-intervals-scikit-learn-is-all-you-need-4b68143a027a

### LINKED ISSUE #2161: Adding bootstrapping functionnality from residuals of a model
It would be interesting to be able to add bootstrapping functionnality from residuals of a model of the time series, this would help : 

1. to compute prediction intervals when the method doesn't support it like for regression model (example here : https://otexts.com/fpp3/prediction-intervals.html) and access the full predictive distribution for these models;
2. when a normal distribution for the residuals is an unreasonable assumption boostrapping is a good alternative (example here : https://otexts.com/fpp3/prediction-intervals.html);
3. when we need to compute aggregate of forecast being compute at lower level (example here : https://otexts.com/fpp3/aggregates.html);
4. generate new time series that are similar to our observed series using blocked bootstrap, this can be used for example to compute bagged forecast in order to improve forecast accuracy (https://otexts.com/fpp3/bootstrap.html);

### PR DESCRIPTION
**Pull request by dennisbader**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #1704, fixes #2161.

### Short Summary
- Adds the first two Conformal Prediction Models: `ConformalNaiveModel`, and `ConformalQRModel` (read more below).
- Adds 3 new quantile interval metrics (plus their aggregated versions):
  - Interval Winkler Score `iws()`, and Mean Interval Winkler Scores `miws()` (time-aggregated) ([source](https://otexts.com/fpp3/distaccuracy.html))
  - Interval Coverage `ic()` (binary if observation is within the quantile interval), and Mean Interval Covarage `mic()` (time-aggregated)
  - Interval Non-Conformity Score for Quantile Regression `incs_qr()`, and Mean ... `mincs_qr()` (time-aggregated) ([source](https://arxiv.org/pdf/1905.03222))
- Adds support for `overlap_end=True` in `ForecastingModel.residuals()`. This computes historical forecasts and residuals that can extend further than the end of the target series. With this, all returned residual values have the same length per forecast (the last residuals will contain missing values, if the forecasts extended further into the future than the end of the target series). 

### Summary

Adds first conformal prediction models to Darts. Conformal models can be applied to any of Darts' global forecasting model, as long as the model has been fitted before. In general the workflow of the models to produce one forecast/prediction is as follows:

- Extract a calibration set: The calibration set for each conformal forecast is automatically extracted from
  the past of your input series relative to the forecast start point. The number of calibration examples
  (forecast errors / non-conformity scores) to consider can be defined at model creation
  with parameter `cal_length`. Note that when using `cal_stride>1`, a longer history is required since
  the calibration examples are generated with stridden historical forecasts.
- Generate historical forecasts on the calibration set (using the forecasting model) with a stride `cal_stride`.
- Compute the errors/non-conformity scores (specific to each conformal model) on these historical forecasts
- Compute the quantile values from the errors / non-conformity scores (using our desired quantiles set at model
  creation with parameter `quantiles`).
- Compute the conformal prediction: Using these quantile values, add calibrated intervals to (or adjust the
  existing intervals of) the forecasting model's predictions.

> Notes:
> - When computing `historical_forecasts()`, `backtest()`, `residuals()`, ... the above is applied for each forecast
> - For **multi-horizon** forecasts, the above is applied for each step in the horizon separately
> - Focus was put on keeping it as efficient as possible using mostly "vectorized" operations

#### Input Support

All added conformal models support the following input (depending on the fitted forecasting model):
- uni/multivariate target series
- past/future/static covariates
- single/multiple series

#### Forecast/Output Support

All models support the following prediction modes:
- **single/multi-horizon** forecasts. For multi-horizon, the calibration process is repeated per step in the forecast horizon.
- **single/mutliple quantile intervals**: It can be any number of quantile intervals as long as they are centered around the median (e.g., `quantiles=[0.05, 0.2, 0.5, 0.8, 0.95]`).
- historical forecasts with expanding or rolling calibration sets with parameter `cal_length` (to make the algorithm adaptive)
- **direct quantile predictions** using `predict_likelihood_parameters=True, num_samples=1` in all prediction methods. 
- **sampled predictions** from these quantile predictions using `num_samples>>1` in all prediction methods.

#### Requirements to use a conformal model:

- Any pre-trained GlobalForecastingModel (global baselines, all regression models, all torch models)
- A long enough calibration set, depending on the forecast horizon `n`. It must be possible to generate at least `n + cal_length` historical forecasts from the calibration input series.

#### Added Algorithms

Added two algorithms each with two symmetry modes:

- `ConformalNaiveModel`: Adds calibrated intervals around the median forecast from the forecasting model.
  - `symmetric=True`: 
     - The lower and upper interval bounds are calibrated by the same magnitude. 
     - Non-conformity scores: uses metric [`ae()`](https://unit8co.github.io/darts/generated_api/darts.metrics.metrics.html#darts.metrics.metrics.ae) (absolute error) to compute the non-conformity scores
  - `symmetric=False`
     - The lower and upper interval bounds are calibrated separately
     - Non-conformity scores: uses metric [`err()`](https://unit8co.github.io/darts/generated_api/darts.metrics.metrics.html#darts.metrics.metrics.err) (error) to compute the non-conformity scores of the upper bounds, an `-err()` for the lower bounds.
- `ConformalQRModel` (Conformalized Quantile Regression, [source](https://arxiv.org/pdf/1905.03222)): Calibrates the quantile predictions from a probabilistic forecasting model.
  - `symmetric=True`: 
     - The lower and upper interval bounds are calibrated by the same magnitude. 
     - Non-conformity scores: uses metric [`incs_qr(symmetric=True)`](https://unit8co.github.io/darts/generated_api/darts.metrics.metrics.html#darts.metrics.metrics.incs_qr) (Quantile Regression Non-Conformity Score) to compute the non-conformity scores
  - `symmetric=False`
     - The lower and upper interval bounds are calibrated separately
     - Non-conformity scores: uses metric [`incs_qr(symmetric=False)`](https://unit8co.github.io/darts/generated_api/darts.metrics.metrics.html#darts.metrics.metrics.incs_qr) (Quantile Regression Non-Conformity Score) to compute the non-conformity scores for the upper and lower bound separately.

### CODE REVIEW

**Review by madtoinou — COMMENTED** [review]
Amazing work @dennis!

Some small comments, mostly documentation.

File: darts/models/forecasting/forecasting_model.py [hunk 1]
```diff
@@ -654,42 +659,60 @@ def historical_forecasts(
...
+        """Generates historical forecasts by simulating predictions at various points in time throughout the history of
+        the provided (potentially multiple) `series`. This process involves retrospectively applying the model to
+        different time steps, as if the forecasts were made in real-time at those specific moments. This allows for an
+        evaluation of the model's performance over the entire duration of the series, providing insights into its
+        predictive accuracy and robustness across different historical periods.
+
+        There are two main modes for this method:
+
+        - Re-training Mode (Default, `retrain=True`): The model is re-trained at each step of the simulation, and
+          generates a forecast using the updated model.
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/forecasting_model.py:678]: 👍

File: darts/tests/models/forecasting/test_historical_forecasts.py [hunk 2]
```diff
@@ -2599,3 +2601,424 @@ def test_sample_weight(self, config):
...
+        pred_lklp = {"num_samples": 1, "predict_likelihood_parameters": True}
+        # compute minimum series length to generate n forecasts
+        icl = 3
+        ocl = 5
+        horizon_ocs = horizon + ocs
+        min_len_val_series = icl + horizon_ocs + int(not overlap_end) * horizon_ocs
+        n_forecasts = 3
+        # get train and val series of that length
+        series_train, series_val = (
+            self.ts_pass_train[:10],
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/tests/models/forecasting/test_historical_forecasts.py]: it doesn't really matter since we don't check for performance (only that given some residuals, we expect some calibrated intervals)

File: darts/metrics/metrics.py [hunk 3]
```diff
@@ -3259,16 +3328,17 @@ def iw(
 ) -> METRIC_OUTPUT_TYPE:
     """Interval Width (IL).
```
Comment (flaessig) [inline darts/metrics/metrics.py:3329]: Why IL?

File: darts/metrics/metrics.py [hunk 4]
```diff
@@ -3280,7 +3350,7 @@ def iw(
         For time series that are overlapping in time without having the same time index, setting `True`
         will consider the values only over their common time interval (intersection in time).
     q_interval
-        The quantile interval(s) to compute the metric on. Must be a tuple (single interval) or sequence tuples
+        The quantile interval(s) to compute the metric on. Must be a tuple (single interval) or sequence of tuples
         (multiple intervals) with elements (low quantile, high quantile).
     q
         Quantiles `q` not supported by this metric; use `q_interval` instead.
```
[the remark this answers is not in the document]
Comment (flaessig) [reply darts/metrics/metrics.py:3356]: would be nice to have a bit more documentation on what `get_quantile_intervals` does

File: darts/models/forecasting/conformal_models.py [hunk 5]
```diff
@@ -0,0 +1,1858 @@
...
+          creation with parameter `quantiles`).
+        - Compute the conformal prediction: Using these quantile values, add calibrated intervals to (or adjust the
+          existing intervals of) the forecasting model's predictions.
+
+        Parameters
+        ----------
+        n
+            Forecast horizon - the number of time steps after the end of the series for which to produce predictions.
+        series
+            A series or sequence of series, representing the history of the target series whose future is to be
```
Comment (flaessig) [inline darts/models/forecasting/conformal_models.py:276]: If I understand correctly, if the calibration series overlaps or is contained in the training set, this may lead to overly optimistic prediction intervals. Should we add this disclaimer?

File: darts/models/forecasting/conformal_models.py [hunk 6]
```diff
@@ -0,0 +1,1858 @@
...
+        **kwargs,
+    ) -> "ConformalModel":
+        """Fit/train the underlying forecasting model on (potentially multiple) series.
+
+        Optionally, one or multiple past and/or future covariates series can be provided as well, depending on the
+        forecasting model used. The number of covariates series must match the number of target series.
+
+        Notes
+        -----
+        Conformal Models do not required calling `fit()`, since they use pre-trained global forecasting models.
```
Comment (madtoinou) [inline darts/models/forecasting/conformal_models.py:194]:
```suggestion
        Conformal Models do not require calling `fit()`, since they use pre-trained global forecasting models.
```

**Review by madtoinou — APPROVED** [review]
It looks really nice, the notebook require a few adjustment/improvement and I think that this is ready to be merged.

There are a few "WIP" keyword in some docstring and in the notebook, let's maybe get ride of them ;)

File: darts/models/forecasting/conformal_models.py (hunk 5 above)
Comment (dennisbader) [reply darts/models/forecasting/conformal_models.py:276]: Sounds good 👍

File: darts/models/forecasting/conformal_models.py [hunk 7]
```diff
@@ -0,0 +1,1858 @@
...
+        - Compute the conformal prediction: Using these quantile values, add calibrated intervals to (or adjust the
+          existing intervals of) the forecasting model's predictions.
+
+        Parameters
+        ----------
+        n
+            Forecast horizon - the number of time steps after the end of the series for which to produce predictions.
+        series
+            A series or sequence of series, representing the history of the target series whose future is to be
+            predicted. Will use the past of this series for calibration.
```
Comment (flaessig) [inline darts/models/forecasting/conformal_models.py:277]:
```suggestion
            predicted. Will use the past of this series for calibration.
            Please note that if the calibration series overlaps with or is contained in the training set, this
            may lead to overly optimistic prediction intervals.
```
